experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp

```diff
@@ -196,6 +229,530 @@ struct PrintOpLowering : public ConversionPattern {
+
+    // Extract upstream state.
+    Value initialState = block->getArgument(0);
+    Value initialUpstreamState = rewriter.create<ExtractValueOp>(
+        op->getLoc(), upstreamInfo.stateType, initialState,
+        rewriter.getIndexArrayAttr({0}));
+
+    // Get first result from upstream.
+    SmallVector<Type> nextResultTypes = {upstreamInfo.stateType,
+                                         rewriter.getI1Type(), elementType};
+    func::CallOp firstNextCall =
+        rewriter.create<func::CallOp>(op->getLoc(), upstreamInfo.nextFunc,
+                                      nextResultTypes, initialUpstreamState);
+
+    // Check for empty upstream.
+    scf::IfOp ifOp = rewriter.create<scf::IfOp>(op->getLoc(), nextResultTypes,
+                                                firstNextCall->getResult(1),
+                                                /*withElseRegion=*/true);
+
+    {
+      // Fill "then" region of ifOp.
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointToStart(&ifOp.getThenRegion().front());
+
+      // Create while loop.
+      SmallVector<Value> whileInputs = {firstNextCall->getResult(0),
+                                        firstNextCall->getResult(2)};
+      SmallVector<Type> whileInputTypes = {
+          firstNextCall->getResult(0).getType(),
+          firstNextCall->getResult(2).getType()};
+      SmallVector<Type> whileResultTypes = {upstreamInfo.stateType, elementType,
+                                            elementType};
+      SmallVector<Location> whileResultLocs = {op->getLoc(), op->getLoc(),
+                                               op->getLoc()};
+      scf::WhileOp whileOp = rewriter.create<scf::WhileOp>(
+          op->getLoc(), whileResultTypes, whileInputs);
+
+      // Fill the "before" region of whileOp.
+      {
+        OpBuilder::InsertionGuard guard(rewriter);
+        Block *before =
+            rewriter.createBlock(&whileOp.getBefore(), {}, whileInputTypes,
+                                 {op->getLoc(), op->getLoc()});
+        Value currentState = before->getArgument(0);
+        Value currentAggregate = before->getArgument(1);
+        func::CallOp nextCall = rewriter.create<func::CallOp>(
+            op->getLoc(), upstreamInfo.nextFunc, nextResultTypes, currentState);
+        rewriter.create<scf::ConditionOp>(op->getLoc(), nextCall->getResult(1),
+                                          ValueRange{nextCall->getResult(0),
+                                                     nextCall->getResult(2),
+                                                     currentAggregate});
+      } // "before" region
+
+      // Fill the "after" region of whileOp.
+      {
+        OpBuilder::InsertionGuard guard(rewriter);
+        Block *after = rewriter.createBlock(&whileOp.getAfter(), {},
+                                            whileResultTypes, whileResultLocs);
+
+        Value currentState = after->getArgument(0);
```

Comment:
Let's do a similar refactoring here first if you don't mind, this will allow seeing what can truly compose at the type level vs what needs Operation* and is limited in its ability to look back into the IR.

Reply:
Not sure I fully understand. I have done refactorings to (1) have a generic part that works on `Operation *` that takes care of a lot of plumbing and (2) extract the op-specific conversion logic into `build*Body` functions, which work on CRTP `Op` types. I guess you are referring to this `Operation *` vs `Op` question, right?